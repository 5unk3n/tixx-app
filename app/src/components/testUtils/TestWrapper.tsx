import { NavigationContainer } from '@react-navigation/native'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import React, { PropsWithChildren } from 'react'
import { PaperProvider } from 'react-native-paper'

import { theme } from '@/theme/defaultTheme'

const queryClient = new QueryClient({
	defaultOptions: {
		queries: {
			retry: false,
			staleTime: Infinity
		}
	}
})

export const TestWrapper = ({ children }: PropsWithChildren) => {
	return (
		<PaperProvider theme={theme.dark}>
			<QueryClientProvider client={queryClient}>
				<NavigationContainer>{children}</NavigationContainer>
			</QueryClientProvider>
		</PaperProvider>
	)
}
