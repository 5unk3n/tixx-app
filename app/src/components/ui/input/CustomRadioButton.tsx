import { styled } from 'nativewind'
import React, { createContext, useContext, ReactNode } from 'react'
import { View, ViewStyle, TextStyle } from 'react-native'
import { TouchableRipple } from 'react-native-paper'

import { useCustomTheme } from '@/hooks/useCustomTheme'

import CustomIcon from '../display/CustomIcon'
import { CustomText } from '../display/CustomText'

interface RadioGroupContextType {
	value: string
	onChange: (value: string) => void
}

const RadioGroupContext = createContext<RadioGroupContextType | null>(null)

interface RadioGroupProps extends RadioGroupContextType {
	style?: ViewStyle
	children: ReactNode
}

function RadioGroup({ children, value, onChange, style }: RadioGroupProps) {
	return (
		<RadioGroupContext.Provider value={{ value, onChange }}>
			<View style={style}>{children}</View>
		</RadioGroupContext.Provider>
	)
}

const iconMap = {
	minimalChecked: 'checkboxMinimalChecked',
	minimalUnchecked: 'checkboxMinimalUnchecked',
	roundedChecked: 'radioCheckedPrimary',
	roundedUnchecked: 'radioUncheckedPrimary'
} as const

interface RadioButtonProps {
	value: string
	label: string
	iconType?: 'minimal' | 'rounded'
	style?: ViewStyle
	textStyle?: TextStyle
}

function RadioButton({
	value,
	label,
	iconType = 'minimal',
	style,
	textStyle
}: RadioButtonProps) {
	const { colors } = useCustomTheme()
	const context = useContext(RadioGroupContext)
	if (!context) {
		throw new Error('RadioButton must be used within a RadioGroup')
	}

	const { value: groupValue, onChange } = context
	const isSelected = value === groupValue

	return (
		<TouchableRipple
			onPress={() => onChange(value)}
			rippleColor={'transparent'}
		>
			<View
				className={`flex-row ${iconType === 'rounded' ? 'p-1' : 'px-7 py-3'}`}
				style={style}
			>
				<CustomText
					className="flex-1"
					variant="headline2MediumLarge"
					style={textStyle}
				>
					{label}
				</CustomText>
				<CustomIcon
					name={
						iconType === 'minimal'
							? isSelected
								? iconMap.minimalChecked
								: iconMap.minimalUnchecked
							: isSelected
								? iconMap.roundedChecked
								: iconMap.roundedUnchecked
					}
					color={colors.primary}
				/>
			</View>
		</TouchableRipple>
	)
}

export const CustomRadioButton = {
	Group: RadioGroup,
	Button: styled(RadioButton, { props: { textStyle: true } })
}
