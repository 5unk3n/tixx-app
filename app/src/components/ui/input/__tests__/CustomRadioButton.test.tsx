import { describe, it, expect, jest } from '@jest/globals'
import { render, screen, fireEvent } from '@testing-library/react-native'
import React from 'react'

import { TestWrapper } from '../../../testUtils/TestWrapper'
import { CustomRadioButton } from '../CustomRadioButton'

describe('CustomRadioButton', () => {
	it('should switch selection state within a context', () => {
		const onChangeMock = jest.fn()
		render(
			<CustomRadioButton.Group value="a" onChange={onChangeMock}>
				<CustomRadioButton.Button value="a" label="옵션A" />
				<CustomRadioButton.Button value="b" label="옵션B" />
			</CustomRadioButton.Group>,
			{ wrapper: TestWrapper }
		)

		fireEvent.press(screen.getByText('옵션B'))
		expect(onChangeMock).toHaveBeenCalledWith('b')
	})

	describe('should render correctly based on props', () => {
		const testCases = [
			{ value: 'a', iconType: 'minimal' }, // Unchecked minimal
			{ value: 'b', iconType: 'minimal' }, // Checked minimal
			{ value: 'c', iconType: 'rounded' }, // Unchecked rounded
			{ value: 'd', iconType: 'rounded' } // Checked rounded
		] as const

		testCases.forEach(({ value, iconType }) => {
			it(`renders correctly for value '${value}' with iconType '${iconType}'`, () => {
				const { toJSON } = render(
					<CustomRadioButton.Group value={'b'} onChange={jest.fn()}>
						<CustomRadioButton.Button
							value={value}
							label={`옵션 ${value}`}
							iconType={iconType}
						/>
					</CustomRadioButton.Group>,
					{ wrapper: TestWrapper }
				)
				expect(toJSON()).toMatchSnapshot()
			})
		})
	})
})
