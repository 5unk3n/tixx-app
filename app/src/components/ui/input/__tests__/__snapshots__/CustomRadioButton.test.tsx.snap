// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomRadioButton should render correctly based on props renders correctly for value 'a' with iconType 'minimal' 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            false,
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "paddingLeft": 28,
                "paddingRight": 28,
              },
              {
                "paddingBottom": 12,
                "paddingTop": 12,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "#FFFFFF",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "Pretendard-Medium",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "letterSpacing": 0,
                    "lineHeight": 22,
                  },
                  [
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                  ],
                ],
              ]
            }
          >
            옵션 a
          </Text>
          <test-file-stub
            color="#F2F862"
            height={26}
            pointerEvents="none"
            width={26}
          />
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomRadioButton should render correctly based on props renders correctly for value 'b' with iconType 'minimal' 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            false,
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "paddingLeft": 28,
                "paddingRight": 28,
              },
              {
                "paddingBottom": 12,
                "paddingTop": 12,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "#FFFFFF",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "Pretendard-Medium",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "letterSpacing": 0,
                    "lineHeight": 22,
                  },
                  [
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                  ],
                ],
              ]
            }
          >
            옵션 b
          </Text>
          <test-file-stub
            color="#F2F862"
            height={26}
            pointerEvents="none"
            width={26}
          />
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomRadioButton should render correctly based on props renders correctly for value 'c' with iconType 'rounded' 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            false,
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "paddingBottom": 4,
                "paddingLeft": 4,
                "paddingRight": 4,
                "paddingTop": 4,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "#FFFFFF",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "Pretendard-Medium",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "letterSpacing": 0,
                    "lineHeight": 22,
                  },
                  [
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                  ],
                ],
              ]
            }
          >
            옵션 c
          </Text>
          <test-file-stub
            color="#F2F862"
            height={26}
            pointerEvents="none"
            width={26}
          />
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`CustomRadioButton should render correctly based on props renders correctly for value 'd' with iconType 'rounded' 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            false,
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flexDirection": "row",
              },
              {
                "paddingBottom": 4,
                "paddingLeft": 4,
                "paddingRight": 4,
                "paddingTop": 4,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "#FFFFFF",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "Pretendard-Medium",
                    "fontSize": 18,
                    "fontWeight": "500",
                    "letterSpacing": 0,
                    "lineHeight": 22,
                  },
                  [
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                  ],
                ],
              ]
            }
          >
            옵션 d
          </Text>
          <test-file-stub
            color="#F2F862"
            height={26}
            pointerEvents="none"
            width={26}
          />
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
